### PR TITLE
Nntaus

### PR DESCRIPTION
**Pull request by violatingcp**
#### PR description:

This pull request  is an update to PUPPI and the NN Tau algorithm to ensure continual full performance of the NN Tau algorithm in all scenarios. The changes consist of : 

1. The use of multiple vertices for the PUPPI computation: 
        a. For the non bitwise emulator Set by https://github.com/violatingcp/cmssw/blob/nntaus/L1Trigger/Phase2L1ParticleFlow/python/l1pfProducer_cfi.py#L25
        b. For the full ctl1 emualtion, its set by : 
        https://github.com/violatingcp/cmssw/blob/nntaus/L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py#L32
and nVtx in the rest of the config above
        c. This algorithm will take the first N vertices of the given vertex collection and treat the charged particles as fromPV for the PUPPI calculation. 
        d. If N is not satisfied, which has been the case for recent vertex algorithms that only output 1 vertex, this algorithm will run its own vertexing and take any additional vertices, up to N, to satisfy the vertex requirement. The vertexing run is just the same as the TDR (Histogram based vertexing). This feature can be shut off in future running
       e. For Tau performance with VBF, we observed 2 vtx was sufficient to gain the efficiency. However, with many algorithms, such as the TDR algorithm, this 2nd vertex is not needed. 

2. The addition of a fix bit precision NN Tau algorithm based on the actual hardware implementation
      a. Set by https://github.com/violatingcp/cmssw/blob/nntaus/L1Trigger/Phase2L1ParticleFlow/python/L1NNTauProducer_cff.py#L16
      b. The algorithm seeds the taus using the deregionzier algorithm, that also has a full emulator
      c. Due to parallel development the current algorithm uses its own dergionizer algorithm, merging with the existing deregionizer code in CMSSW is possible,  but is currently planned for the next implementation of this algorithm
      d. To decrease the background rate, particularly at high pT, we have added a new option selection : 
            https://github.com/violatingcp/cmssw/blob/nntaus/DataFormats/L1TParticleFlow/interface/PFTau.h#L51-L53
           This selection is not currently endorsed for the simple fact that the gain at low tau pT is relatively small.

#### PR validation:

Here is a link to slides that discuss in detail performance of this setup:
https://www.dropbox.com/s/077g8ym0p8dollr/PCH_TauValid_10_06.pdf?dl=0

### CODE REVIEW

**Comment by gpetruc** [comment]
Cecile/Enrique/Vladimir: please hold off merging this before we discuss it:
changes to Puppi were only validated for taus and not for jets & met yet,
it would break the emulator agreement with firmware for layer 1, and at a
first look it seems to also be commenting out muon inputs to PF.

File: L1Trigger/Phase2L1ParticleFlow/python/L1NNTauProducer_cff.py
Comment (mcepeda) [inline L1Trigger/Phase2L1ParticleFlow/python/L1NNTauProducer_cff.py:29]: int32 instead of uint32

**Comment by violatingcp** [comment]
> 

The muon comments are a mistake, I started from 3.4.6 following Dylan's suggestion. I agree that these changes having nothing to do with muons. Also, I am personally hesitant to switch to multiple vertices unless it is absolutely needed. However, I think adding the multiple vertice functionality would be useful going forward.

**Comment by cbotta** [comment]
@gpetruc: 
The DPG would like to test/study the performance of the updated NNTau emulator, without and with the multiple-vertex functionality. The NN Tau emulator performance have been presented by the developers back in Summer, but never validated by the DPG using the central software. We understand that the multiple vertices PUPPI is validated only for Taus, and for the moment we will just consider it as an "alternative" option. Our understanding is that this PR is  just  adding a "multiple vertex PUPPI" option, and it's not changing the default PUPPI (i.e. the one used by all other PUPPI based algos). Isn't it the case?

**Comment by violatingcp** [comment]
@cbotta at the moment 2 vertices is turned on. If we are to run this, then this would be a change of ALL PUPPI based algorithms. There is really no way to run 2 PUPPIs, especially one just for taus. We have done some studies in the past and found that multiple vertices has a small impact. However, I have a feeling that with all of the other changes that have incurred, we will not need to run multiple vertices for the NNTaus to reach optimal performance.

**Comment by gpetruc** [comment]
@violatingcp @cbotta While it won't be realistic, in CMSSW we can temporary run two copies of Layer1, one with the single PV (and emulator agreement) and a copy with NPV = 2 for the taus.
E.g. in l1ctLayer1_cff  one can keep the default configuration as is (nVtx = 1) and towards the end do
````
      l1ctLayer1Barrel2PV = l1ctLayer1Barrel.clone(nVtx = 2) 
      l1ctLayer1HGCal2PV = l1ctLayer1HGCal.clone(nVtx = 2) 
      l1ctLayer12PV = l1ctLayer1.clone(pfProducers = [
        cms.InputTag("l1ctLayer1Barrel2PV"),
        cms.InputTag("l1ctLayer1HGCal2PV"),
        cms.InputTag("l1ctLayer1HGCalNoTK"),
        cms.InputTag("l1ctLayer1HF")
    ])
````
an add it to `l1ctLayer1Task`.

To me this seems the best option in the short term to allow you to do studies with the new taus without affecting the rest.
Then in the slightly longer term we can figure out if there's a GTT vertexing emulator that can deliver multiple PVs, test the jets+MET performance and eventually implement it in the full layer1 firmware.

I'll have a few more comments on the code later today when I have the time to look at it more carefully.

**Comment by cbotta** [comment]
@gpetruc thanks, it's a good proposal. Will you make the changes? For what concerns "we can figure out if there's a GTT vertexing emulator that can deliver multiple PVs", my understanding is that the current emulator already does that, and that emulated vertex collection PUPPI is currently using, has up to 10 vertices. We discussed this in the GTT last year, when HLT asked for this feature, i.e. L1 to send more vertices.

**Comment by cbotta** [comment]
@gpetruc @violatingcp I think that the best would be 2 https://github.com/violatingcp/cmssw/blob/nntaus/L1Trigger/Phase2L1ParticleFlow/python/L1NNTauProducer_cff.py#L10 one that runs over the std PUPPI and one that runs over the 2tvx PUPPI, and in both cases I think we want the HW set to true by default (given it turns on the fix bit precision algo)

**Comment by gpetruc** [comment]
@cbotta: I can make the change as a PR to @violatingcp's repo but it's cleaner if Phil does it directly together with the other fixes

**Comment by violatingcp** [comment]
@cbotta  and @gpetruc  I can  do that, I will also fix the standalone muon issue that Giovanni pointed out.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc]: This won't work as expected, because it will only use the tracks in the regions considered by this module, so you'll get a barrel vertex made with just barrel tracks and an endcap vertex made just with endcap tracks.
You should loop instead on the sectors in event_.decoded.track, which would give you all the tracks (and you can skip the isFiducial check later)

That said, I would very much prefer if we could just read a collection of multiple PVs from GTT rather than resurrect this hacked in-house vertex finding.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc]: Revert this

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:381]: Can you put all this changes inside of an if nVtx > 1? 
So that we keep the existing functionality for nVtx = 1

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp]: this indentation issue means you probably forgot to do `scram b code-format`

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp]: Don't add a new copy of the functions, just change the existing ones to take a vector of PVObjs.
That should also make the changes easier to review.
I'll review the rest of the changes to these files once this is done

**Review by gpetruc — CHANGES_REQUESTED** [review]
I've put some comments on things that need changes.

I didn't look at the old L1 PFProducer modifications since that part is obsolete anyway and we will just remove it from CMSSW in one of the next PRs

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc]: agreed that we shouldn't use this I added it as an option in the config that is currently set to false. I don't see any drop in efficiency. I also don't observe a drop in efficiency with 1 vertex.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc]: fixed

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:381]: fixed

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp]: ran this update

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.cpp]: updated

**Comment by violatingcp** [comment]
I have updated the tau tagger now with the option of running two vertices. I have also set the HW by default, along with the l1ct inputs. Updates are here : 
https://github.com/violatingcp/cmssw/blob/nntaus/L1Trigger/Phase2L1ParticleFlow/python/L1NNTauProducer_cff.py#L19-L21

Additionally, I have reverted the default PUPPI back to 1 vertex, further I have added a task of the tau which does 2 vertices here : 
https://github.com/violatingcp/cmssw/blob/nntaus/L1Trigger/Phase2L1ParticleFlow/python/L1NNTauProducer_cff.py#L38-L56

Lastly, I have checked the performance with and without 2Vtx and I don't see a significant difference in efficiency.

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:376]: If addExtraVtx_ is true, then you're no longer guaranteed that the first vertex is the one already added in `event_.pvs.push_back(hwpv);` above, so by skipping the first of the z0s you may be skipping one vertex and duplicating another one.
So, I think you need

     if (addExtraVtx_) {
         event_.pvs.clear();
         doVertexings(z0s);
     }
    for (unsigned int i = (addExtraVtx_ ? 0 : 1); i < z0s.size(); ++i) {

That said, if you get ok performance of the 2 vertex case by just changing the configuration in `L1Trigger/L1TTrackMatch/python/L1TkPrimaryVertexProducer_cfi.py` then I would prefer to remove this entirely

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.h
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.h:150]: you forgot to update also this file changing the function instead of adding the new one

**Comment by cbotta** [comment]
@violatingcp : on your comment "Lastly, I have checked the performance with and without 2Vtx and I don't see a significant difference in efficiency." I'm confused, I thought that this code refers to your performance plots presented here https://indico.cern.ch/event/1045568/contributions/4392201/attachments/2262258/3840147/PCH_TauValid_10_06.pdf. I was therefore expecting a difference in efficiency same as the green vs violet in slide 9. Could you clarify please? thanks!

**Comment by violatingcp** [comment]
@cbotta Apologies if I was not clear. What I mean to say is that with the current vertexing algorithm, I don't see the drop in efficiency that was present in previous versions. Here is the current performance I see ###
Screen Shot 2022-01-20 at 7 49 51 AM

File: L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/plugins/L1TCorrelatorLayer1Producer.cc:376]: Agreed, this was really a hacky fix when the vertexer was only outputting 1 vertex. Updated.

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.h
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/src/newfirmware/puppi/linpuppi_ref.h:150]: fixed

**Comment by cbotta** [comment]
@violatingcp ok thanks for explaning. Still, I don't understand what causes the difference between purple and green. PUPPI should now be using the "emulated" vertices collection, which still comes from the FastHisto algorithm, as in the past. Moreover which signal samples are you using? (If it is VBF Htautau, then the purple efficiency is the one we have been showing since the TDR time).

**Comment by violatingcp** [comment]
The vertexing algorithm I used was from an old version of L1TkPrimaryVertex. I am not sure if there was an issue with that version that is causing this.

**Comment by cbotta** [comment]
@violatingcp ok, so, it's not clear and we need to investigate. Since the TDR time the NNPuppiAlgo uses the simulated PUPPI, with the simulated FastHisto (L1TkPrimaryVertex) in input: https://github.com/cms-l1t-offline/cmssw/blob/l1t-phase2-v3.3.45/L1Trigger/Phase2L1ParticleFlow/python/l1ParticleFlow_cff.py#L39-L40 and we have always seen ~70% plateau efficiency for VBFHtautau. Therefore we have now to understand where the improvement in efficiency comes from while using the CTL1 emulation, and vertex emulation, in input. @gpetruc @mcepeda can this be merged? so that we can perform these studies?

File: L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py]: vtxRes no longer needed or used anywhere, so remove from here and the rest of the file

File: L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py:193]: same for `etraVtx`, which is also misspelled

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/TauNNIdHW.cc
Comment (gpetruc) [inline L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/TauNNIdHW.cc]: Why all these are in jetmet?

I think the best would be to move:
 - the common part which is not nn tau specific into a new directory `newfirmware/common/hls4ml`
 - the NN tau specific part under `newfirmware/taus/nn`

**Review by gpetruc — CHANGES_REQUESTED** [review]
Almost there.

**Comment by gpetruc** [comment]
@cbotta since now the rest is cosmetics (which should be fixed before merging, but don't affect functionality) I'll run a quick test of this version now

**Comment by gpetruc** [comment]
I tested the PR and, provided the code changes are done, it's fine for me: it makes no change at all in default PF+Puppi.

**Comment by cbotta** [comment]
perfect! @violatingcp could you apply the requested code changes? @mcepeda once this PR is merged, we want to include in the Menu ntuplizer the collection produced by L1NNTauProducerPuppi, and the one produced by L1NNTauProducerPuppi2Vtx. In both cases we want to store the same variables as before, plus the new ones passMass and passLooseNNMass. Thanks!

File: L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/TauNNIdHW.cc
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/src/newfirmware/jetmet/TauNNIdHW.cc]: moved to taus, and organized things to more like an sw framework. They were just there b/c that is where I got it to work.

File: L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py
Comment (violatingcp) [reply L1Trigger/Phase2L1ParticleFlow/python/l1ctLayer1_cff.py:193]: cleaned up.

**Comment by gpetruc** [comment]
I'm ok with the updated version that Phil pushed to the github 88b4e83

**Comment by epalencia** [comment]
Tagged as **l1t-phase2-v3.4.13**

**Comment by mcepeda** [comment]
For completeness, adding vtxRes to the configuration for l1ctLayer1 - and a couple other tiny changes -  the code runs and I can get the nntaus (2vtx) in the menu ntuples. It looks like a trivial change (and I do not see the variable being used in the producer?), but before going ahead I would like to clear the change with @gpetruc, @violatingcp , @cbotta.

**Pull request merged**
